pkg/agent/plugin/workloadattestor/docker/docker.go

```diff
@@ -182,52 +160,115 @@ func (p *Plugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi
+		candidate, ok := finder.FindContainerID(cgroup.GroupPath)
+		if !ok {
+			continue
+		}
 
-	p.containerIDFinder, err = cgroup.NewContainerIDFinder(matchers)
-	if err != nil {
-		return nil, err
-	}
+		hasDockerEntries = true
 
-	p.findContainerID = p.containerIDFinder.FindContainerID
+		switch {
+		case containerID == "":
+			// This is the first container ID found so far.
+			containerID = candidate
+		case containerID != candidate:
+			// More than one container ID found in the cgroups.
+			return "", fmt.Errorf("workloadattestor/docker: multiple container IDs found in cgroups (%s, %s)",
+				containerID, candidate)
+		}
+	}
 
-	return &spi.ConfigureResponse{}, nil
+	switch {
+	case !hasDockerEntries:
+		// Not a docker workload. Since it is expected that non-docker workloads will call the
+		// workload API, it is fine to return a response without any selectors.
+		return "", nil
+	case containerID == "":
+		// The "finder" found a container ID, but it was blank. This is a
+		// defensive measure against bad matcher patterns and shouldn't
+		// be possible with the default finder.
+		return "", fmt.Errorf("workloadattestor/docker: a pattern matched, but no container id was found")
+	default:
+		return containerID, nil
+	}
 }
 
-func (*Plugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
-	return &spi.GetPluginInfoResponse{}, nil
+type legacyContainerIDFinder struct {
+	log                  hclog.Logger
+	cgroupPrefix         string
+	cgroupContainerIndex int
 }
 
-func (p *Plugin) legacyExtractID(cgroupPath string) (string, bool) {
-	if !strings.HasPrefix(cgroupPath, p.cgroupPrefix) {
+func (f *legacyContainerIDFinder) FindContainerID(cgroupPath string) (string, bool) {
+	if !strings.HasPrefix(cgroupPath, f.cgroupPrefix) {
 		return "", false
 	}
 
 	parts := strings.Split(cgroupPath, "/")
 
-	if len(parts) <= p.cgroupContainerIndex {
-		p.log.Warn("Docker entry found, but is missing the container id", telemetry.CGroupPath, cgroupPath)
+	if len(parts) <= f.cgroupContainerIndex {
+		f.log.Warn("Docker entry found, but is missing the container id", telemetry.CGroupPath, cgroupPath)
 		return "", false
 	}
 
-	return parts[p.cgroupContainerIndex], true
+	return parts[f.cgroupContainerIndex], true
+}
+
+// dockerCGroupRE matches cgroup paths that have the following properties.
+// 1) `\bdocker\b` the whole word docker
+// 2) `.+` followed by one or more characters (which will start on a word boundary due to #1)
+// 3) `\b([[:xdigit:]][64])\b` followed by a 64 hex-character container id on word boundary
+//
+// The "docker" prefix and 64-hex character container id can be anywhere in the path. The only
+// requirement is that the docker prefix comes before the id.
+var dockerCGroupRE = regexp.MustCompile(`\bdocker\b.+\b([[:xdigit:]]{64})\b`)
+
+type defaultContainerIDFinder struct{}
+
+func (f *defaultContainerIDFinder) FindContainerID(cgroupPath string) (string, bool) {
```

Comment:
I think that a small comment about this function would be good. We can point that `dockerCGroupRE` is used here and it would be good to explain the meaning of an empty string result and the boolean returned.